internal/db/content.go

```diff
@@ -62,40 +87,146 @@ func (c *ContentEncoder) Encode(content []byte) ([]byte, error) {
-		output = []byte("")
+	if tmpOutput == nil {
+		return []byte(""), nil
 	}
 
+	output := make([]byte, len(tmpOutput))
+	copy(output, tmpOutput)
 	return output, nil
 }
 
+func (c *ContentEncoder) Close() error {
+	return c.writer.Close()
+}
+
 type ContentDecoder struct {
+	buffer *bytes.Buffer
 	reader *s2.Reader
 }
 
 func NewContentDecoder() *ContentDecoder {
+	var buffer bytes.Buffer
 	reader := s2.NewReader(nil)
 
 	return &ContentDecoder{
+		buffer: &buffer,
 		reader: reader,
 	}
 }
 
-func (c *ContentDecoder) Decoder(encoded []byte) ([]byte, error) {
-	c.reader.Reset(bytes.NewBuffer(encoded))
-	output, err := io.ReadAll(c.reader)
+func (c *ContentDecoder) Decode(encoded EncodedContent) (DecodedContent, error) {
+	c.buffer.Reset()
+	c.reader.Reset(bytes.NewReader(encoded))
+
+	_, err := io.Copy(c.buffer, c.reader)
 	if err != nil {
 		return nil, err
 	}
 
+	output := make([]byte, c.buffer.Len())
+	copy(output, c.buffer.Bytes())
 	return output, nil
 }
 
+type ContentLookup struct {
+	cache    *ristretto.Cache
+	decoders *puddle.Pool[*ContentDecoder]
+}
+
+func NewContentLookup() (*ContentLookup, error) {
+	cache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: 100_000,
+		MaxCost:     1_000 * MB,
+		BufferItems: 64,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	constructor := func(context.Context) (*ContentDecoder, error) {
+		return NewContentDecoder(), nil
+	}
+
+	decoders, err := puddle.NewPool(&puddle.Config[*ContentDecoder]{Constructor: constructor, MaxSize: DecoderPoolSize})
+	if err != nil {
+		return nil, err
+	}
+
+	return &ContentLookup{
+		cache:    cache,
+		decoders: decoders,
+	}, nil
+}
+
+func (cl *ContentLookup) Lookup(ctx context.Context, tx pgx.Tx, hashes map[Hash]bool) (map[Hash]DecodedContent, error) {
+	var notFound []Hash
+	contents := make(map[Hash]DecodedContent, len(hashes))
+
+	decoder, err := cl.decoders.Acquire(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot acquire content decoder: %w", err)
+	}
+	defer decoder.Release()
+
+	for hash, isEncoded := range hashes {
+		value, found := cl.cache.Get(hash.Hex())
+		if found {
+			if isEncoded {
+				decoded, err := decoder.Value().Decode(value.(EncodedContent))
+				if err != nil {
+					return nil, fmt.Errorf("cannot decode value from cache %v: %w", hash.Hex(), err)
+				}
+				contents[hash] = decoded
+			} else {
+				contents[hash] = value.(DecodedContent)
+			}
+		} else {
+			notFound = append(notFound, hash)
+		}
+	}
+
+	if len(notFound) > 0 {
+		rows, err := tx.Query(ctx, `
+			SELECT (hash).h1, (hash).h2, bytes
+			FROM dl.contents
+			WHERE hash = ANY($1::hash[])
+		`, notFound)
+		if err != nil {
+			return nil, fmt.Errorf("lookup missing hash contents: %w", err)
+		}
+
+		for rows.Next() {
+			var hash Hash
+			var value []byte
+
+			err = rows.Scan(&hash.H1, &hash.H2, &value)
+			if err != nil {
+				return nil, fmt.Errorf("content lookup scan: %w", err)
+			}
+
+			cl.cache.Set(hash.Hex(), value, int64(len(value)))
```

Comment:
I almost commented on race conditions / write-through caching, but then I remembered the thing that's really important here is that everything is content addressable!

Just dropping the comment here in case you think we can make that clearer in any way.